XXX-staged-require-with-lifecycle-hooks.md

```diff
@@ -0,0 +1,192 @@
+accept multiple function arguments and dynamically identifying which is a
+callback that needs to be instrumented can be unreliable. Any form of queueing,
+even simply putting a callback in an array to be called elsewhere, will break
+continuation linking and requires manual intervention to maintain that linkage
+which is not always possible for dynamic instrumentation.
+
+Most APM providers try/catch constantly because the existing solutions for
+tracking continuation lose context regularly. We've done the best we can with
+monkey-patching, but we've hit a wall and have been pushing for things like
+`async-hooks` to improve the situation. However, `async-hooks` only solves part
+of the problem--linking the callsite to the callback--but context on what was
+happening still requires monkey-patching, and there are plenty of situations
+where context-loss is not the hard part.
+
+### Performance
+
+Monkey-patching is prone to performance issues. The vast majority requires
+wrapping things in closures. More complex instrumentation can often trigger
+deoptimization. Sometimes heavy runtime checking is required with constant
+dynamic function generation and reassignment, which can put a lot of undue
+stress on the optimizer and garbage collector.
+
+If, for example, one could intercept the module loader and apply an AST
+transform before compilation, there's a great deal that could be gained.
+Most Run-time checking would simply disappear. Most closures would disappear.
+Context loss would disappear because the transformer could take advantage of
+lexical scope to describe linkage directly at callsites.
+
+### Flexibility
+
+Many things are just not possible to any reasonable capacity relying purely on
+monkey-patching. Often things which could provide valuable coverage or tracing
+insight are simply not exposed.
+
+An example of a difficult thing to apply tracing hooks to is the timing from
+when a socket first enters JS in `net.js` to when it switches to a TLS socket.
+One could listen for the `secureConnection` event, but there are several
+failure paths between the `TLSSocket` constructor and the event handler being
+triggered, which could result in context loss. One might also think to patch
+`TLSSocket.prototype._init`, however that will be called by both servers and
+clients, which are difficult to differentiate from in this case. To correctly
+identify and correlate a net socket to a corresponding TLS upgrade takes
+extensive monkey-patching with a lot of run-time checks in a major hot-path.
+With an AST transform, a single line of code could be inserted
+[before the `TLSSocket` construction][1] in the `Server` constructor.
+
+### ES Modules
+
+A big issue on the horizon is that, if ES Modules are adopted, modules become
+immutable at build time. This means that the current monkey-patching approach
+will not work anymore. TC39 has already suggested using AST transformation
```

Comment:
It came up in one-on-one discussion with @bmeck at one point. Not sure if that is doc'd somewhere.